File: go/core/catalog.go
package core

import (
	"context"
	"encoding/json"

	"github.com/traq-markdown-parser/sdk/go/binding"
)

type catalog struct {
	Groups  []groupDefinition  `json:"groups"`
	Plugins []pluginDefinition `json:"plugins"`
	Rules   []struct {
		Name  *string `json:"name"`
		Phase string  `json:"phase"`
	} `json:"rules"`
	Presets []struct {
		Description string `json:"description"`
		Plugins     []int  `json:"plugins"`
		Order       []int  `json:"order"`
	} `json:"presets"`
}

func displayName(name *string) string {
	if name == nil {
		return ""
	}
	return *name
}
func (r *Runtime) loadCatalog(ctx context.Context) error {
	// The artifact was checked against this generated catalog on instantiation.
	var spec catalog
	if err := json.Unmarshal(binding.CatalogJSON, &spec); err != nil {
		return err
	}
	groups := make([]*PluginGroup, len(spec.Groups))
	for index, item := range spec.Groups {
		group := NewPluginGroup(displayName(item.Name))
		if item.Parent != nil {
			group.parent = groups[*item.Parent]
		}
		groups[index] = group
	}
	rules := make([]*Rule, len(spec.Rules))
	for index, item := range spec.Rules {
		rules[index] = &Rule{runtime: r, index: index, name: displayName(item.Name), phase: item.Phase}
	}
	plugins := make([]*Plugin, len(spec.Plugins))
	for index, item := range spec.Plugins {
		plugin := NewPlugin(displayName(item.Name))
		plugin.text = item.Text
		plugin.providerRuntime = r
		if item.Group != nil {
			plugin.group = groups[*item.Group]
		}
		for _, rule := range item.Rules {
			plugin.rules = append(plugin.rules, rules[rule])
		}
		plugin.frozen = true
		plugins[index] = plugin
	}
	presets := make([]*Grammar, len(spec.Presets))
	for index, item := range spec.Presets {
		if err := ctx.Err(); err != nil {
			return err
		}
		b := r.Builder()
		for _, plugin := range item.Plugins {
			if err := b.Add(plugins[plugin]); err != nil {
				return err
			}
		}
		b.definition.order = item.Order
		snapshot := b.definition.copy()
		recipe, err := json.Marshal(snapshot.composition())
		if err != nil {
			return err
		}
		presets[index] = r.newGrammar(snapshot, recipe, item.Description)
	}
	r.exportCatalog(plugins, presets)
	return nil
}

File: typescript/src/core/parser/catalog.ts
import type { Owner, Catalog, CatalogTree, Grammar } from "./types.js";
import { Plugin as Declaration, PluginGroup } from "../definitions/index.js";
import { Plugin, makePlugin, makeRule } from "./plugin.js";
import { data } from "./identity.js";
export function loadCatalog(runtime: Owner, catalog: Catalog) {
  const groups: PluginGroup[] = [];
  for (const definition of catalog.groups) {
    let group =
      definition.parent === null
        ? new PluginGroup(definition.name)
        : groups[definition.parent].group(definition.name);
    groups.push(group);
  }
  const rules = catalog.rules.map((definition, index) =>
    makeRule(runtime, index, definition),
  );
  const plugins = catalog.plugins.map((definition) => {
    const plugin = new Plugin(
      definition.group === null
        ? new Declaration(definition.name)
        : groups[definition.group].new(definition.name),
    );
    const state = data(plugin, "plugin");
    return makePlugin({
      ...state,
      name: definition.name,
      frozen: true,
      rules: definition.rules.map((index) => rules[index]),
      text: definition.text.map((index) => ({ runtime, index })),
    });
  });
  function tree<T>(
    value: number | CatalogTree,
    leaf: (index: number) => T,
  ): T | { readonly [key: string]: T | object } {
    if (typeof value === "number") return leaf(value);
    return Object.freeze(
      Object.fromEntries(
        Object.entries(value).map(([name, child]) => [name, tree(child, leaf)]),
      ),
    );
  }
  const presets = new Map<number, Grammar>();
  return {
    plugins: tree(catalog.exports.plugins, (index) => plugins[index]),
    presets: tree(catalog.exports.presets, (index) => {
      if (!presets.has(index)) {
        const definition = catalog.presets[index];
        presets.set(
          index,
          runtime.preset(
            {
              plugins: definition.plugins.map((index) =>
                data(plugins[index], "plugin"),
              ),
              order: definition.order,
            },
            definition,
          ),
        );
      }
      return presets.get(index)!;
    }),
  };
}
